Accordion component

mperrotti opened the pull request:
#### Related issues
For the upcoming work on event create

#### Description
A component to handle cases where we hide and show panels of related content. Each panel has a label that is always visible and summarizes the content that will be shown by clicking the label.

#### Screenshots (if applicable)
![screen shot 2017-02-13 at 3 06 19 pm](https://cloud.githubusercontent.com/assets/2313998/22900935/1ad6b27e-f1fe-11e6-8046-e42cf8a8c938.png)

unjust commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+				<li
+					className='list-item'
+					key={i}
+				>
```
nitpick - should this just be on one line? also can line 51 be broken up on a few lines since its so long?

mperrotti replied:
I think it's easier to read on 2 lines, but I could see moving the last angle bracket up to line 57 :)
I'd prefer just keeping line 51 to 1 line. All it's doing is making a string to use for aria.

unjust commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+								className: accordionPanel.props.className,
+								isOpen: this.state.activePanels.includes(i),
+								onTriggerClick: (e) => this._onClickTrigger(e, i),
+							}
```
should `i` become a prop of the Accordion element, like a data attribute, in order to identify which panel is being open and closed in `_onClickTrigger`? Maybe using the index from iterating in map is fine, but this way you could just have `onTriggerClick: this._onClickTrigger`

mperrotti replied:
That could work - I'll try it out

unjust commented on src/accordionPanelGroup.test.jsx:
```diff
@@ -0,0 +1,15 @@
+		expect(accordionPanelGroupNode).not.toBeNull();
+	});
+
+});
```
a couple tests to add would be to simulate clicks on the accordion - make sure the handler gets called and that its in the expected open or closed state. Also check that the accordion panel group state gets updated.

mmcgahan replied:
You check the coverage report when you run `yarn test` to make sure that all the new components have 100% test coverage.

unjust commented on assets/scss/_accordion.scss:
```diff
@@ -0,0 +1,37 @@
+	overflow: hidden;
+	transition: height $A_duration $A_curve-default;
+}
+
```
these can be added to swarm-animation (which has its own animation constants). I don't want to get in the way of this getting merged. we need to think of a way to get this project to pull in animation.scss from swarm-animate pretty soon.

eilinora replied:
@unjust not 100% sure what swarm-animation is but you can checkout how we include WMC & sassquatch in mup-web. might be helpful https://github.com/meetup/mup-web/blob/master/src/assets/scss/main.scss#L11

mperrotti replied:
@unjust - 100% agreed that this stuff should live in `swarm-animation`, but I didn't think it was ready to be used yet.
If we're ready to use `swarm-animation` - let me know and we'll do a follow-up to get that into `meetup-web-components`

unjust commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+						aria-labelledby={`label-${name}`}
+						aria-hidden={!this.state.open}
+						className={classNames.content}
+						ref={(el) => { this.animator = el; }}>
```
what is `this.animator`? reading about refs in React now...

mperrotti replied:
I could go either way. The word "container" sounds a little clearer

mmcgahan replied:
I'd recommend "wrapper" over "container" because "container" has a different meaning in Redux-land (Container == Smart component connected to app state)

akdetrick commented on src/accordionPanel.story.jsx:
```diff
@@ -0,0 +1,26 @@
+
```
I'd recommend creating one story for all accordion features, `accordion.story.jsx`.

You can add a story for individual panel usage, and another for usage in a group.

mperrotti replied:
That makes sense to me

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+				activePanelsArr.push(i);
+			}
+			return activePanelsArr;
+		});
```
I don't think this `map` call is doing what you expect it to. `this.accordionPanels` will end up being an array that is the length of `this.props.accordionPanels`, where each element is a reference to the same `activePanelsArr` array. You actually never _read_ from `this.accordionPanels` (it gets re-assigned at the top of `renderAccordionPanels`), so I think you are intending to just `reduce` `this.props.accordionPanels` into an `activePanelsArr`

```js
const activePanelsArr = this.props.accodionPanels.reduce((panels, panel, i) => {
  if (panel.props.isOpen) {
    panels.push(i);
  }
  return panels;
}, []);
```

That said, it looks like you're actually storing an array of _indexes_ into the active panels, not the panels themselves, so the `activePanelsArr` naming is confusing. I think the show/hide logic is getting a little convoluted, though, so I'll leave a separate comment about options for cleaning it up.

mperrotti replied:
Never used/heard of `reduce`, thanks!

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+		});
+
+		this.state = {
+			activePanels: activePanelsArr || []
```
`activePanelsArr` is always truthy here, even when it's empty.

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+
+		this.setState({
+			activePanels: activePanelArr
+		});
```
this function is very difficult to follow - there's a lot of index matching and searching and filtering, but it's not clear what the end result is. I'm pretty sure that this whole component can be significantly simplified by breaking down the problem, but I'm finding it difficult to know where to start.

Maybe we can pair on this sometime soon so you can walk me through it

mperrotti replied:
I'd love to pair on this @mmcgahan - I'll look for a slot on your calendar

eilinora commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	_updateHeight() {
+		const content = ReactDOM.findDOMNode(this.content);
```
Using ReactDOM.findDOMNode() is considered a bad in the context of a component is considered a bad practice. Since `this.content` is already a reference to the DOM element you should be able to access the elements height. You want to go with https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect instead.

mperrotti replied:
Good catch. I'll go back and change this :)

mperrotti replied:
You're right - Lori had similar feedback.

eilinora commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+		const height = this.props.isOpen ? content.clientHeight : 0;
+		const cssHeight = `${height}px`;
+
+		if (animator.style.height === cssHeight) {
```
What is this checking? :)

mperrotti replied:
Forgot to delete this from a different way I was trying to handle this. If you want more details, follow-up on Slack :)

eilinora commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+			isOpen, // eslint-disable-line no-unused-vars
+			name,
+			onTriggerClick,
+			panelContent,
```
It looks like not all props are listed in PropTypes, can we add them in?

mperrotti replied:
Yes

eilinora commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+					</Chunk>
+				</FlexItem>
+
+				{triggerIconShape &&
```
As an alternative to passing in all the props for declaring the icon, instead let's pass in the <Icon/> element with the proper declarations set instead. This allows for Icon to be less coupled to this component as well

mperrotti replied:
Good idea. Will do

eilinora commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	/**
+	 * Document this for PR
```
we should probably document this as it's going into master

mperrotti replied:
Yes, forgot to go back and update this before submitting the PR

eilinora commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	/**
+	 * Document this for PR
```
we should probably document this as it's going into master

mperrotti replied:
Yes

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+		}
+
+		if (!this.props.multiSelectable) {
+			_.pullAll(activePanelArr, filtered);
```
I am strongly against importing `lodash` as a dependency - it is a very bloated library and 90% redundant with native JS array methods. In this case, I'm not exactly sure why you're pulling out the `filtered` indexes from `activePanelArr`, but I think this is the same as the more-explicit `.filter` alternative

```js
const actuallyActivePanelArr = activePanelArr.filter(panelIndex =>
  filtered.indexOf(panelIndex) === -1
);
```

Although there's some confusing mapping from `activePanelArr` to `filtered` and back into `activePanelArr`, so I'm not sure what the goal state is.

mperrotti replied:
Filter makes more sense, but you're right - I'm doing this in kind of a weird way.

eilinora commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+				role='tabList'
+				aria-multiselectable={multiSelectable}
+				className={classNames}
+				{...other}>
```
nitpick but easier to read when `>` is on the next line when we do this kind of code style of indented props

mperrotti replied:
I've seen it done both ways. If having it on a separate line is our preferred style, I'll move it down

eilinora commented on src/accordionPanel.test.jsx:
```diff
@@ -0,0 +1,15 @@
+
+describe('AccordionPanel', function() {
+
+	it('exists', function() {
```
Tests are very light for this :)

mperrotti replied:
Yes, going to write some tests based on Ivy's feedback

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	renderAccordionPanels() {
+		const toCamelCase = (str) => str.slice(0,20).replace(/(?:^\w|[A-Z]|\b\w)/g, (ltr, idx) => idx === 0 ? ltr.toLowerCase() : ltr.toUpperCase()).replace(/\s+/g, '');
```
We have a [`toCamelCase` string helper](https://github.com/meetup/meetup-web-platform/blob/master/src/util/stringUtils.js#L16) in the platform repo

mperrotti replied:
Thanks, I'll use that instead

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+
+	renderAccordionPanels() {
+		const toCamelCase = (str) => str.slice(0,20).replace(/(?:^\w|[A-Z]|\b\w)/g, (ltr, idx) => idx === 0 ? ltr.toLowerCase() : ltr.toUpperCase()).replace(/\s+/g, '');
+		this.accordionPanels = this.props.accordionPanels.map((accordionPanel, i) => {
```
looks like you can set a `const` rather than a property on`this` - `this.accordionPanels` is not consumed anywhere else.

mperrotti replied:
I see. Will do

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,129 @@
+								name: toCamelCase(accordionPanel.props.triggerLabel),
+								className: accordionPanel.props.className,
+								isOpen: this.state.activePanels.includes(i),
+								onTriggerClick: (e) => this._onClickTrigger(e, i),
```
keep naming consistent `onClickTrigger` -> `_onClickTrigger`

mperrotti replied:
👍

unjust commented on assets/scss/_accordion.scss:
```diff
@@ -0,0 +1,37 @@
+	transition: transform $A_duration $A_curve-default;
+
+	.accordionPanel--active & {
+		-webkit-transform: rotate(180deg);
```
bourbon has one, but it is being deprecated.

mperrotti replied:
Something for @akdetrick to consider for [swarm-sasstools](https://github.com/meetup/swarm-sasstools)

akdetrick replied:
We will not be updating to newer versions of Bourbon for this reason. Once `swarm-sasstools` is in all of our new platform repos, I'm considering moving away from Bourbon to something like [postcss-loader](https://github.com/postcss/postcss-loader) to handle prefixing.

mperrotti commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+	_updateHeight() {
+		const content = ReactDOM.findDOMNode(this.content);
+		const animator = ReactDOM.findDOMNode(this.animator);
+		const height = this.props.isOpen ? content.clientHeight : 0;
```
Clever. I like it and I'm changing it :)

mperrotti commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+			return;
+		}
+
+		animator.style.height = cssHeight;
```
Excellent point. Will fix

mperrotti commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+			triggerIconShapeActive,
+			triggerIconSize,
+			triggerIconAlign,
+			triggerLabel,
```
@mmcgahan: that's how I originally had this, but then I can't set default props for the key values in `icon`. For example: if I want to use a default `size`, but change `shape`, values for everything but `shape` get thrown out.

If you know a way around this, let me know and I'll gladly change it

mmcgahan replied:
you can set defaults in the `constructor`

```js
const triggerDefaults = {
  a: 'b',
  c: 'd',
};
// ...
constructor(props) {
  props.trigger = { ...props.trigger, ...triggerDefaults };
  // ...
}

mperrotti commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+		return(
+			<Flex
+				className={classNames.accordionPanel}
+				rowReverse={triggerIconAlign == 'left' ? 'atAll' : null}
```
Yup 👍

mperrotti commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+							className='accordionPanel-label display--block span--100'
+							onClick={onTriggerClick || this._handleToggle}>
+								{triggerLabel}
+						</button>
```
Good point—should be pretty straightforward.

mperrotti commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,167 @@
+			triggerIconSize,
+			triggerIconAlign,
+			triggerLabel,
+			animated,
```
Yes

mperrotti replied:
Agreed

mmcgahan commented:
bump. what's the status on this? is there a deadline?

mperrotti commented:
@mmcgahan - no hard deadline, but Ivy will be using this component in Event Create.
I've addressed all feedback, I just have to write tests. This is my first time writing tests, so Ivy agreed to pair with me tomorrow afternoon.

akdetrick commented:
Going with individual props for icon options so defaults can be clearly defined (and documented):
![screen shot 2017-03-21 at 10 36 29 pm](https://cloud.githubusercontent.com/assets/231252/24179692/09f897c4-0e87-11e7-938f-027be2fa4ebe.png)

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+		super(props);
+
+		this.state = {
+			open: this.props.isOpen
```
keep naming consistent - `isOpen: this.props.isOpen`

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+		}
+
+		this.setState({
+			height: `${newState * this.content.getBoundingClientRect().height}px`,
```
Since this height calculation is made 3 times, I'd recommend making it a function that takes `isOpen` as an argument

```js
height: getHeight(isOpen),
```

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+	 * Updates state to toggle `AccordionPanel` open and closed
+	 */
+	_handleToggle(){
+		const newState = !this.state.open;
```
keep naming consistent

```js
const  isOpen = !this.state.isOpen;
```

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+				className={classNames.accordionPanel}
+				rowReverse={iconAlign === 'left' && 'all'}
+				{...other}
+				>
```
align closing `>` with opening `<Flex`

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+							aria-expanded={this.state.open}
+							aria-selected={this.state.open}
+							className={classNames.trigger}
+							onClick={this._handleToggle}>
```
align closing `>` with opening `<button` - that will make it clear you have an extra indent on `{label}`

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+						aria-labelledby={`label-${ariaId}`}
+						aria-hidden={!this.state.open}
+						className={classNames.content}
+						style={{height: this.state.height}}>
```
align closing `>` with opening `<Chunk`

mmcgahan replied:
cool story: if you pass Number values to `style` properties, it will automatically append `px`, which means you don't have to have those ugly template strings in your `this.state.height` assignments - just use the actual number

mmcgahan replied:
I just realized that it looks like `this.state.height` is always derived directly from the current value of `this.state.open`, so why store the value of `height` in state at all? Could you just calculate it here each time you render?

akdetrick replied:
seemed like a good idea, but got a helpful warning:
```
Warning: a `div` tag (owner: `Chunk`) was passed a numeric string value for CSS property `height` (value: `184`) which will be treated as a unitless number in a future version of React.
```

mmcgahan replied:
ohhh. okay I will update my knowledge.

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+						style={{height: this.state.height}}>
+						<div
+							className='accordionPanel-content'
+							ref={(div) => { this.content = div; }}>
```
nitpick but `content` is _extremely vague_ for something that is referenced in separate functions - I recommend `contentEl` or something that indicates what kind of thing it is.

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+					className='accordionPanel-icon'
+					onClick={this._handleToggle}
+					shrink
+					>
```
align closing `>` with opening `<FlexItem`

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,198 @@
+					>
+					<Icon
+						shape={this.getIconShape()}
+						size={iconSize} />
```
align closing `>` with opening `<Icon`, or just make the whole opening tag inline - it's short

```jsx
<Icon shape={this.getIconShape()} size={iconSize} />
```

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,132 @@
+	constructor(props) {
+		super(props);
+		this.accordionPanels = this.props.accordionPanels.map((accordionPanel, i) => {
+			return this.clonePanels(accordionPanel, i, accordionPanel.props.isOpen);
```
since `clonePanels` takes a single panel as an argument, could you name it `clonePanel`?

Also, since the third argument is always a property of the first argument, you don't need to pass it in separately, which means that you can simplify this to

```js
this.props.accordionPanels.map(this.clonePanel);
```

However, you should `bind` `this.closePanel` _before_ calling it in this way.

akdetrick replied:
yeah, changing this to `clonePanel` makes sense. 

The `isOpen` argument is either provided by the props of the `AccordionPanel`, or provided by some logic in `AccordionPanelGroup`. After giving some refactoring a shot, I feel like keeping the argument makes it easier to understand when we're relying on the panel props vs. explicitly setting `isOpen`.

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,132 @@
+	 * Keeps track of the clicked panel in order to support AccordionPanelGroups that only
+	 * have one panel open at a time.
+	 */
+	setClickedPanel(clickedPanel) {
```
Thinking about this a bit more, and I'm wondering if you should just _always_ use `setState({ clickedPanel })`, but ignore the value when `multiSelectable` is `false` rather than trying to be smart about setting `this.setClickedPanel` to be false and then checking whether it's false before calling it... essentially make the method calling consistent, by putting all the logic into figuring out what to _do_ with the information it provides

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,132 @@
+	 * @returns {Array} `AccordionPanel` components with the correct value for `isOpen` prop
+	 */
+	renderAccordionPanels() {
+		this.accordionPanels = this.accordionPanels.map((accordionPanel, i) => {
```
pretty sure this could also be simplified to `this.accordionPanels = this.accordionPanels.map(this.clonePanels)`

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,132 @@
+				aria-multiselectable={multiSelectable}
+				className={classNames}
+				{...other}
+				>
```
align closing `>` with opening `<ul`

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,132 @@
+						</li>
+						)
+					)
+				}
```
indentation is a little weird here - simplify

```jsx
this.renderAccordionPanels().map((panel, key) => (
  <li key={key} className='list-item'>{panel}</li>
)
```

nlaz commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,132 @@
+			const isOpen = accordionPanel.props.isOpen;
+
+			/*
+			 *if ( !this.props.multiSelectable && this.state.clickedPanel === accordionPanel ) {
```
Is this needed anymore?

akdetrick commented on src/accordion.test.jsx:
```diff
@@ -0,0 +1,185 @@
+		openPanel = null;
+	});
+
+	it(`has the class ${ACTIVEPANEL_CLASS} when state is open`, function() {
```
Given that the panel content visibility is based on `height` manipulation rather than CSS class toggling, we'll have some trouble checking this in tests.

Our thinking was that covering state changes and the `ACTIVEPANEL_CLASS` would be adequate for alerting us when content visibility might break

akdetrick commented:
Bump; how are we feeling about this one in its current state?

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,203 @@
+	 * Updates state to toggle `AccordionPanel` open and closed
+	 */
+	_handleToggle(){
+		const newState = !this.state.isOpen;
```
can this be `isOpen` to avoid a new variable alias?

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,203 @@
+	 */
+	_handleToggle(){
+		const newState = !this.state.isOpen;
+		const { setClickedPanel } = this.props;
```
nitpick, but since this var is only used once in an `if` test, I'd skip the destructuring and just write `if (this.props.setClickedPanel)`

mmcgahan commented on src/AccordionPanel.jsx:
```diff
@@ -0,0 +1,203 @@
+	/**
+	 * @returns {undefined}
+	 *
+	 * Sets height of `AccordionPanel` to be appear open or closed when mounting
```
style nitpick - JSDoc function description usually goes _above_ the `@` tags

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,115 @@
+	 */
+	clonePanel(accordionPanel, i, isOpen) {
+		const panelProps = {
+			key: i,
```
trivial, but feel free to name the index argument `key` so that you don't have to alias it.

mmcgahan commented on src/AccordionPanelGroup.jsx:
```diff
@@ -0,0 +1,115 @@
+				{
+					this.renderAccordionPanels().map((panel, i) => (
+							<li key={i} className='list-item'> {panel} </li>
+						)
```
lines 90 and 91 have an extra indent - the close paren on 91 should go on the same line as the one on 92 since their corresponding open parens are both on line 89

akdetrick commented:
This should be good, as long as we feel ok about https://github.com/meetup/meetup-web-components/pull/79#discussion-diff-107768714R148
